ml-agents/mlagents/trainers/stats.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from typing import List, Dict, NamedTuple
+import numpy as np
+import abc
+
+from mlagents.tf_utils import tf
+
+
+class StatsWriter(abc.ABC):
+    """
+    A StatsWriter abstract class. A StatsWriter takes in a category, key, scalar value, and step
+    and writes it out by some method.
+    """
+
+    @abc.abstractmethod
+    def write_stats(self, category: str, key: str, value: float, step: int) -> None:
+        pass
+
+    @abc.abstractmethod
+    def write_text(self, category: str, text: str, step: int) -> None:
+        pass
+
+
+class TensorboardWriter(StatsWriter):
+    def __init__(self):
+        self.summary_writers = {}
+
+    def write_stats(self, category: str, key: str, value: float, step: int) -> None:
+        if category not in self.summary_writers:
+            self.summary_writers[category] = tf.summary.FileWriter(category)
+        summary = tf.Summary()
+        summary.value.add(tag="{}".format(key), simple_value=value)
+        self.summary_writers[category].add_summary(summary, step)
+        self.summary_writers[category].flush()
+
+    def write_text(self, category: str, text: str, step: int) -> None:
+        if category not in self.summary_writers:
+            self.summary_writers[category] = tf.summary.FileWriter(category)
+        self.summary_writers[category].add_summary(text, step)
+
+
+class StatsSummary(NamedTuple):
+    mean: float
+    std: float
+    num: int
+
+
+class StatsReporter:
+    def __init__(self, writers: List[StatsWriter]):
+        """
+        Generic StatsReporter. A category is the broadest type of storage (would
```

Comment:
Rather than having a global StatsReporter and passing in `category` for every `add_stat`, why not have a StatsReporter for every category?

Reply:
That's possible, but there still needs to be a way to access commonly used StatsReporters, e.g. the AgentProcessor and the Trainer will record in the same category. What about if we had the global class store a Dict of StatsReporters, and each module would grab one and use it?

Reply:
How about making `StatsReporter.writers` and `StatsReporter.stats_dict` static members, and accepting a "category" as the only argument of the constructor?